forch/faucet_state_collector.py

```diff
@@ -823,6 +823,23 @@ def process_port_learn(self, timestamp, name, port, mac, ip_addr):
                 devices_placement = DevicePlacement(switch=name, port=port, connected=True)
                 self._placement_callback(mac, devices_placement)
 
+    @_dump_states
+    # pylint: disable=too-many-arguments
+    def process_port_expire(self, timestamp, name, port, mac):
+        """process port learn event"""
+        with self.lock:
+            LOGGER.info('Learned entry %s at %s:%s expired.', mac, name, port)
+            if self.learned_macs:
+                entry = self.learned_macs.pop(mac, None)
+                if not entry:
+                    LOGGER.debug('Entry %s doesnt exist in learned macs dict', mac)
+            if name in self.switch_states:
+                learned_macs = self.switch_states[name][LEARNED_MACS]
+                if mac in learned_macs:
+                    learned_macs.remove(mac)
+                else:
+                    LOGGER.debug('Entry %s doesnt exist in learned macs set', mac)
```

Comment:
include name in message
should be warning?  When might this happen?

Reply:
We get a learning expire event for each point at which it was learned. I used to debug behavior so made it debug. Need not be a warning. Can remove completely if needed.

Reply:
You didn't answer the question -- when does this case (man not in learned_macs) happen?  Seems like it shouldn't happen, which is why I'm suggesting a warning -- else it's pointing to some other underlying problem in the algorithm?

What about the case where it expires on one switch but not the others?

Reply:
It happens if the mac is learnt at multiple switches. So you get multiple L2_EXPIRE events. The first one deletes it. The second one would not find the entry. I dont think we should be raising a warning for that.

Reply:
What about the case where it expires on one switch but not the others?

Reply:
yup. you are right. Fixed code to remove entry from set only when the last entry is removed.